pkg/cni/types.go

```diff
@@ -0,0 +1,104 @@
+	"net/http"
+	"time"
+
+	"github.com/containernetworking/cni/pkg/types/current"
+)
+
+type cniRequestFunc func(request *PodRequest) ([]byte, error)
+
+// Explicit type for CNI commands the server handles
+type command string
+
+// PodRequest represents a request for ADD / DEL / CHECK for a Pod
+type PodRequest struct {
+	// The CNI command of the operation
+	Command command
+	// kubernetes namespace name
+	PodNamespace string
+	// kubernetes pod name
+	PodName string
+	// kubernetes pod UID
+	PodUID string
+	// kubernetes container ID
+	SandboxID string
+	// CNI container ID
+	ContainerID string
+	// kernel network namespace path
+	Netns string
+	// Interface name to be configured
+	IfName string
+	// CNI conf obtained from stdin conf
+	CNIConf *types.NetConf
+	// Timestamp when the request was started
+	timestamp time.Time
+	// ctx is a context tracking this request's lifetime
+	ctx context.Context
+	// cancel should be called to cancel this request
+	cancel context.CancelFunc
+	// kubeclient has the kubernetes API client
+	kubeclient *k8sclient.ClientInfo
+	// exec execs
+	exec invoke.Exec
+}
+
+// Request sent to the Server by the multus-shim
+type Request struct {
+	// CNI environment variables, like CNI_COMMAND and CNI_NETNS
+	Env map[string]string `json:"env,omitempty"`
+	// CNI configuration passed via stdin to the CNI plugin
+	Config []byte `json:"config,omitempty"`
+}
+
+// KubeAPIAuth contains information necessary to create a Kube API client
+type KubeAPIAuth struct {
+	// Kubeconfig is the path to a kubeconfig
+	Kubeconfig string `json:"kubeconfig,omitempty"`
+	// KubeAPIServer is the URL of a Kubernetes API server (not required if kubeconfig is given)
+	KubeAPIServer string `json:"kube-api-server,omitempty"`
+	// KubeAPIToken is a Kubernetes API token (not required if kubeconfig is given)
+	KubeAPIToken string `json:"kube-api-token,omitempty"`
+	// KubeCAData is the Base64-ed Kubernetes API CA certificate data (not required if kubeconfig is given)
+	KubeCAData string `json:"kube-ca-data,omitempty"`
+}
+
+// Server represents an HTTP server listening to a unix socket. It will handle
+// the CNI shim requests issued when a pod is added / removed.
+type Server struct {
+	http.Server
+	requestFunc       cniRequestFunc
+	rundir            string
+	useOVSExternalIDs int32
+	kubeAuth          *KubeAPIAuth
+	kubeclient        *k8sclient.ClientInfo
+	exec              invoke.Exec
+}
+
+// CNIAdd is the command representing add operation for a new pod
+const CNIAdd command = "ADD"
+
+// CNIUpdate is the command representing update operation for an existing pod
+const CNIUpdate command = "UPDATE"
+
+// CNIDel is the command representing delete operation on a pod that is to be torn down
+const CNIDel command = "DEL"
+
+// CNICheck is the command representing check operation on a pod
+const CNICheck command = "CHECK"
+
+// Plugin represents the connection between the CNI shim and server.
+type Plugin struct {
```

Comment:
Why do you create structure for that? Is there any detailed plan to enhance it?

Reply:
I think you're right; this can be dropped. I have no plans to extend this struct.

Reply:
> Using the encapsulated socket path seems simpler, more readable, and uses less boilerplate code.

I don't think so. Currently functions, CmdAdd/CmdDel/CmdCheck/DoCNI, does not refer `p *Plugin` often, actually.

CmdAdd/CmdDel/CmdCheck: refer once (p.DoCNI)
DoCNI: refer once (p.SocketPath)

So I suppose we could decapsulate it and pass socket as argument to CmdAdd/Del/Check/DoCNI.

Reply:
It's true that its not used often, but being able to simplify 4 calls is better than not doing it.

Furthermore, the caller code (located on the shim) below is simpler (and less):
```golang 
	p := cni.Plugin{SocketPath: cni.SocketPath(*runDir)}
	skel.PluginMain(
		p.CmdAdd,
		p.CmdCheck,
		p.CmdDel,
		cniversion.All, "meta-plugin that delegates to other CNI plugins")
```

Than what is required following your suggestion:
```golang
	skel.PluginMain(
		func(args *skel.CmdArgs) error {
			result, err := cni.CmdAdd(cni.SocketPath(*runDir))
			if err != nil {
				return err
			}
			return result.Print()
		},
		func(args *skel.CmdArgs) error {
			return cni.CmdCheck(cni.SocketPath(*runDir))
		},
		func(args *skel.CmdArgs) error { return cni.CmdDel(cni.SocketPath(*runDir)) },
		cniversion.All, "meta-plugin that delegates to other CNI plugins")
```